vdr/validators.go

```diff
@@ -74,40 +87,191 @@ func (v verificationMethodValidator) verifyThumbprint(method *did.VerificationMe
+}
+
+func (m managedServiceValidator) Validate(document did.Document) error {
+	// make sure that it resolves when if it's a reference
+	var resolvedEndpoint any
+	var err error
+	// Cache resolved DID documents because most of the time all (compound) services will refer to the same DID document in all service references.
+	cache := make(map[string]*did.Document, 0)
+	for _, service := range document.Service {
+		switch se := service.ServiceEndpoint.(type) {
+		case string:
+			resolvedEndpoint, err = m.resolveOrReturnEndpoint(se, cache)
+		case map[string]string:
+			knownKeys := make(map[string]bool, len(se))
+			resolvedCompoundEndpoint := make(map[string]any, len(se)) // don't know if returned type is string or another map
+			for name, endpoint := range se {
+				if _, exists := knownKeys[name]; exists {
+					err = errors.New("duplicate service key in compound service")
+					break
+				}
+				knownKeys[name] = true
+				if resolvedEndpoint, err = m.resolveOrReturnEndpoint(endpoint, cache); err != nil {
+					break
+				}
+				resolvedCompoundEndpoint[name] = resolvedEndpoint
 			}
+			resolvedEndpoint = resolvedCompoundEndpoint
+		default:
+			err = errors.New("invalid service format")
+		}
+		if err != nil {
+			return invalidServiceError{err}
 		}
+
+		// specific service.Type need additional validation
+		if err = serviceTypeValidation(service.Type, resolvedEndpoint); err != nil {
+			return invalidServiceError{fmt.Errorf("%s: %w", service.Type, err)}
+		}
+	}
+	return nil
+}
+
+func (m managedServiceValidator) resolveOrReturnEndpoint(serviceEndpoint string, cache map[string]*did.Document) (any, error) {
+	// make sure that it resolves if it is a reference
+	if didservice.IsServiceReference(serviceEndpoint) {
+		serviceURI, err := ssi.ParseURI(serviceEndpoint)
 		if err != nil {
-			return fmt.Errorf("invalid service: %w", err)
+			return nil, err
 		}
-		knownServiceTypes[method.Type] = true
+		if err = didservice.ValidateServiceReference(*serviceURI); err != nil {
+			return nil, err
+		}
+		resolvedService, err := m.serviceResolver.ResolveEx(*serviceURI, 0, didservice.DefaultMaxServiceReferenceDepth, cache)
+		if err != nil {
+			return nil, err
+		}
+		return resolvedService.ServiceEndpoint, nil
+	}
+	return serviceEndpoint, nil
+}
+
+func serviceTypeValidation(sType string, endpoint any) error {
+	switch sType {
+	case "NutsComm":
+		return validateNutsCommEndpoint(endpoint)
+	case "node-contact-info":
+		return validateNodeContactInfo(endpoint)
+	default:
+		// no extra type-based validation
+		return nil
+	}
+}
+
+func validateNutsCommEndpoint(endpoint any) error {
+	// RFC015 $3.2: NutsComm rules
+	endpointStr, ok := endpoint.(string)
+	if !ok {
+		return errors.New("endpoint not a string")
+	}
+	if URL, err := url.Parse(endpointStr); err != nil {
+		return err
+	} else if URL.Scheme != "grpc" {
+		return errors.New("scheme must be grpc")
+	}
+	return nil
+}
+
+func validateNodeContactInfo(endpoint any) error {
```

Comment:
This is more complicated than necessarry, RFC006 only states that `email` must be present, and others (name, website, etc) _may_ be present. Doesn't state other keys can't be present. Nothing should break when other stuff is there as well.

I'd say just assert `email` is present, and just take the rest for granted.